src/main.rs

```diff
@@ -1,178 +1,155 @@
-
-    let authorized_request = match env::var("CLOUDFLARE_APITOKEN") {
-        Ok(val) => {
-            let mut bearer = "Bearer ".to_owned();
-            bearer.push_str(&val);
-            request.header("Authorization", bearer.to_owned())
-        }
-        Err(_e) =>  {
-            let cloudflare_apikey = env_var("CLOUDFLARE_APIKEY");
-            let cloudflare_email = env_var("CLOUDFLARE_EMAIL");
-            request
-                .header("X-Auth-Key", cloudflare_apikey.to_owned())
-                .header("X-Auth-Email", cloudflare_email.to_owned())
-        }
-    };
-
-    let response_json: Value = authorized_request
-        .send()
-        .unwrap()
-        .json()
-        .unwrap();
-
-    let success = response_json
-        .as_object()
-        .unwrap()
-        .get("success")
-        .unwrap()
-        .as_bool()
-        .unwrap();
-    if !success {
-        return Err(format!("Request not successful: {}", response_json));
-    }
-
-    Ok(response_json)
-}
-
-fn get_current_ip() -> Result<String, ()> {
-    let gdns_addr = (NS1_GOOGLE_COM_IP_ADDR)
-        .parse()
-        .expect("Couldn't get Google DNS Socket Addr");
-    let conn = UdpClientConnection::new(gdns_addr).expect("Couldn't open DNS UDP Connection");
-    let client = SyncClient::new(conn);
-
-    let name = domain::Name::new();
-    let name = name
-        .append_label("o-o")
-        .unwrap()
-        .append_label("myaddr")
-        .unwrap()
-        .append_label("l")
-        .unwrap()
-        .append_label("google")
-        .unwrap()
-        .append_label("com")
-        .unwrap();
-    let response = client.query(&name, DNSClass::IN, RecordType::TXT).unwrap();
-
-    let record = &response.answers()[0];
-    match record.rdata() {
-        &RData::TXT(ref txt) => {
-            let val = txt.txt_data();
-            return Ok(String::from_utf8(val[0].to_vec()).unwrap());
-        }
-        _ => return Err(()),
-    }
+use std::io::prelude::Write;
+
+use std::net::IpAddr;
+
+use public_ip::{http, Version};
+
+use anyhow::Result;
+
+use api::Cli;
+use clap::Parser;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+	let cli = Cli::parse();
+
+	pretty_env_logger::formatted_builder()
+		.filter_level(cli.verbose.log_level_filter())
+		.init();
+
+	log::debug!("Rquested records to update: {:#?}", cli.records);
+
+	let (public_ipv4, public_ipv6) = tokio::join!(
+		public_ip::addr_with(http::ALL, Version::V4),
+		public_ip::addr_with(public_ip::ALL, Version::V6)
+	);
+
+	if (None, None) == (public_ipv6, public_ipv4) {
+		anyhow::bail!("Could not determine your current public IP address.")
+	}
+
+	if let Some(ipv4) = public_ipv4 {
+		log::info!("{}", ipv4);
+	}
+	if let Some(ipv6) = public_ipv6 {
+		log::info!("{}", ipv6);
+	}
+
+	let credentials: Credentials = if let Some(token) = cli.token {
+		Credentials::UserAuthToken { token }
+	} else if let (Some(key), Some(email)) = (cli.key, cli.email) {
+		Credentials::UserAuthKey { email, key }
```

Comment:
related to the question about clap+deprecation, if I'm not missing another mechanism, does it make sense to warn the user here?

Reply:
Yep, it looks like there is an open issue for clap, but it is not yet implemented: clap-rs/clap#3321

So this is probably the place to warn the user :+1: